### ISSUE TITLE
Query Dates before 1970

### ISSUE THREAD

**Issue by nkabir** [juji-io/datalevin#94]
Thank you for this software! I'm eager to incorporate it into my projects. To understand how pieces fit together, I've been implementing the database from https://github.com/jonase/learndatalogtoday and working through the examples.

I encountered unexpected behavior with the dates below. Apologies if this is covered somewhere but I was unable to find an explanation. I know how difficult it is to write comprehensive documentation.

```
user=> (d/transact! conn [{:db/id -1 :person/name "Bob Smith" :person/born #inst "1969-01-01"}])

user=> (d/q '[:find ?e ?born :where [?e :person/born ?born]] db)
#{[3 #inst "1972-12-19T00:00:00.000-00:00"]
  [4 #inst "1970-01-01T00:00:00.000-00:00"]}
```

The query does not return dates before 1970.

**Comment by den1k** [comment]
You need to pass in a function to filter dates. I suggest you post this in https://clojurians.slack.com/ under #datalog #datascript #datalevin and let the people there help you.

**Comment by nkabir** [comment]
Thank you. I will ask there as well but wanted to clarify here since it's surprising behavior. I need to pass a filter function to get *all* the dates? Filters are typically used to restrict returned sets rather than expand them. Is this behavior unique to dates or does it apply to other data types

**Comment by nkabir** [comment]
Here's the schema I'm using. I tried to adapt the source from https://github.com/jonase/learndatalogtoday

```
{:movie/title
 {:db/valueType :db.type/string
  :db/unique :db.unique/identity
  :db/cardinality :db.cardinality/one}

 :movie/year
 {:db/valueType :db.type/long
  :db/cardinality :db.cardinality/one}

 :movie/director
 {:db/valueType :db.type/ref
  :db/cardinality :db.cardinality/many}

 :movie/sequel
 {:db/valueType :db.type/ref
  :db/cardinality :db.cardinality/one}

 :movie/cast
 {:db/valueType :db.type/ref
  :db/cardinality :db.cardinality/many}

 :person/name
 {:db/valueType :db.type/string
  :db/cardinality :db.cardinality/one}

 :person/born
 {:db/valueType :db.type/instant
  :db/cardinality :db.cardinality/one}

 :person/death
 {:db/valueType :db.type/instant
  :db/cardinality :db.cardinality/one}

 :movie/trivia
 {:db/valueType :db.type/string
  :db/cardinality :db.cardinality/many}
}
```

**Comment by den1k** [comment]
this seems to be a bug. here's some repro code:

```clojure
(defonce conn
  (d/create-conn
    "data/instant-test"
    {:person/born {:db/valueType :db.type/instant}}))

(d/transact! conn [{:person/born #inst "1969-01-01"}
                   {:person/born #inst "1971-01-01"}])
;; => {:datoms-transacted 2}
(d/q '[:find [?born ...] :where [?e :person/born ?born]] @conn)
;; [#inst"1971-01-01T00:00:00.000-00:00"]

(d/datoms @conn :ave)

;;=> [#datalevin/Datom[2 :person/born #inst"1971-01-01T00:00:00.000-00:00"]
;     #datalevin/Datom[1 :person/born #inst"1969-01-01T00:00:00.000-00:00"]]
```
@nkabir have you checked if this is also the case with datascript?
@huahaiy do you know what this could be?

**Comment by nkabir** [comment]
> @nkabir have you checked if this is also the case with datascript?

I have not set up Clojurescript/DataScript yet. I'll look into the setup process. Do you know any sites like https://replit.com/ that may have it pre-configured? The docs at DataScript are also fairly light https://github.com/tonsky/datascript/wiki/Getting-started

**Comment by huahaiy** [comment]
I am open to a PR that changes the current implemenation of `:instant` at https://github.com/juji-io/datalevin/blob/master/src/datalevin/bits.cljc#L547, so that it is similar to `timestamp` of Postgres https://www.postgresql.org/docs/13/datatype-datetime.html

**Comment by nkabir** [comment]
Thanks for the pointer, @huahaiy. I will investigate.

**Comment by nkabir** [comment]
I am running Ubuntu 20.04

I forked datalevin and cloned it locally. 
I checked out tag 0.5.28
I ran `lein deps` and `lein compile`
I then ran `lein test` and generated the [following error](https://drive.google.com/file/d/1wtZucuykg7lBguEPeXVrHvjHfhjHIJ0Q/view?usp=sharing):

```
Testing datalevin.client-test
Syntax error (InaccessibleObjectException) compiling at (/tmp/form-init10858281779123311091.clj:1:74).
Unable to make field long java.nio.Buffer.address accessible: module java.base does not "opens java.nio" to unnamed module @5274766b
```

I am fairly new to Clojure so I'm not sure how to follow `unnamed module`.

**Comment by den1k** [comment]
What JDK version are you running? I’ve had issues with 16. 14 usually
works.

**Comment by nkabir** [comment]
Clojure 1.10.3
OpenJDK 64-Bit Server VM 17.0.1+12

**Comment by nkabir** [comment]
I tried

```
openjdk 11.0.12 2021-07-20
OpenJDK Runtime Environment 18.9 (build 11.0.12+7)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.12+7, mixed mode)
```
But that may be too old.

```
Syntax error (UnsupportedClassVersionError) compiling at (datalevin/protocol.clj:1:1).
datalevin/io/ByteBufferInputStream has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 55.0
```
I was hoping to [use an LTS release](https://www.oracle.com/java/technologies/java-se-support-roadmap.html) 

I use SDKMan to manage my versions. Does anyone know if any of these candidates is compatible with Datalevin?

```
================================================================================
Available Java Versions for Linux 64bit
================================================================================
 Vendor        | Use | Version      | Dist    | Status     | Identifier
--------------------------------------------------------------------------------
 Corretto      |     | 17.0.2.8.1   | amzn    |            | 17.0.2.8.1-amzn     
               |     | 17.0.0.35.1  | amzn    |            | 17.0.0.35.1-amzn    
               |     | 11.0.14.10.1 | amzn    |            | 11.0.14.10.1-amzn   
               |     | 11.0.14.9.1  | amzn    |            | 11.0.14.9.1-amzn    
               |     | 11.0.12.7.1  | amzn    |            | 11.0.12.7.1-amzn    
               |     | 8.322.06.2   | amzn    |            | 8.322.06.2-amzn     
 Dragonwell    |     | 11.0.12.8    | albba   |            | 11.0.12.8-albba     
               |     | 8.8.9        | albba   |            | 8.8.9-albba         
 Gluon         |     | 22.0.0.3.r17 | gln     |            | 22.0.0.3.r17-gln    
               |     | 22.0.0.3.r11 | gln     |            | 22.0.0.3.r11-gln    
 GraalVM       |     | 22.0.0.2.r17 | grl     |            | 22.0.0.2.r17-grl    
               |     | 22.0.0.2.r11 | grl     |            | 22.0.0.2.r11-grl    
               |     | 21.3.1.r17   | grl     |            | 21.3.1.r17-grl      
               |     | 21.3.1.r11   | grl     |            | 21.3.1.r11-grl      
               |     | 21.3.1.r8    | grl     |            | 21.3.1.r8-grl       
               |     | 21.3.0.r17   | grl     |            | 21.3.0.r17-grl      
               |     | 21.3.0.r11   | grl     |            | 21.3.0.r11-grl      
               |     | 21.2.0.r16   | grl     |            | 21.2.0.r16-grl      
               |     | 21.2.0.r11   | grl     |            | 21.2.0.r11-grl      
               |     | 21.2.0.r8    | grl     |            | 21.2.0.r8-grl       
               |     | 21.1.0.r8    | grl     |            | 21.1.0.r8-grl       
               |     | 20.3.5.r11   | grl     |            | 20.3.5.r11-grl      
               |     | 20.3.4.r11   | grl     |            | 20.3.4.r11-grl      
               |     | 20.3.3.r11   | grl     |            | 20.3.3.r11-grl      
               |     | 20.3.3.r8    | grl     |            | 20.3.3.r8-grl       
               |     | 20.3.2.r8    | grl     |            | 20.3.2.r8-grl       
               |     | 19.3.6.r11   | grl     |            | 19.3.6.r11-grl      
               |     | 19.3.6.r8    | grl     |            | 19.3.6.r8-grl       
 Java.net      |     | 19.ea.9      | open    |            | 19.ea.9-open        
               |     | 19.ea.3.lm   | open    |            | 19.ea.3.lm-open     
               |     | 19.ea.1.pma  | open    |            | 19.ea.1.pma-open    
               |     | 18.ea.35     | open    |            | 18.ea.35-open       
               |     | 17.0.2       | open    |            | 17.0.2-open         
               | >>> | 11.0.12      | open    | installed  | 11.0.12-open        
               |     | 11.0.11      | open    |            | 11.0.11-open        
               |     | 11.0.10      | open    |            | 11.0.10-open        
               |     | 11.0.2       | open    |            | 11.0.2-open         
               |     | 8.0.302      | open    |            | 8.0.302-open        
               |     | 8.0.292      | open    |            | 8.0.292-open        
               |     | 8.0.282      | open    |            | 8.0.282-open        
               |     | 8.0.265      | open    |            | 8.0.265-open        
 Liberica      |     | 17.0.2.fx    | librca  |            | 17.0.2.fx-librca    
               |     | 17.0.2       | librca  |            | 17.0.2-librca       
               |     | 11.0.14.fx   | librca  |            | 11.0.14.fx-librca   
               |     | 11.0.14      | librca  |            | 11.0.14-librca      
               |     | 8.0.322.fx   | librca  |            | 8.0.322.fx-librca   
               |     | 8.0.322      | librca  |            | 8.0.322-librca      
 Liberica NIK  |     | 22.0.0.r17   | nik     |            | 22.0.0.r17-nik      
               |     | 22.0.0.r11   | nik     |            | 22.0.0.r11-nik      
               |     | 21.3.1.r17   | nik     |            | 21.3.1.r17-nik      
               |     | 21.3.1.r11   | nik     |            | 21.3.1.r11-nik      
               |     | 21.3.0.r17   | nik     |            | 21.3.0.r17-nik      
               |     | 21.3.0.r11   | nik     |            | 21.3.0.r11-nik      
               |     | 21.2         | nik     |            | 21.2-nik            
               |     | 21.1         | nik     |            | 21.1-nik            
               |     | 21.0.0.2     | nik     |            | 21.0.0.2-nik        
 Mandrel       |     | 21.3.0.0     | mandrel |            | 21.3.0.0-mandrel    
               |     | 21.2.0.2     | mandrel |            | 21.2.0.2-mandrel    
               |     | 20.3.3.0     | mandrel |            | 20.3.3.0-mandrel    
 Microsoft     |     | 17.0.2       | ms      |            | 17.0.2-ms           
               |     | 11.0.14      | ms      |            | 11.0.14-ms          
 Oracle        |     | 17.0.2       | oracle  |            | 17.0.2-oracle       
 SapMachine    |     | 17.0.2       | sapmchn |            | 17.0.2-sapmchn      
               |     | 11.0.14      | sapmchn |            | 11.0.14-sapmchn     
               |     | 11.0.14.1    | sapmchn |            | 11.0.14.1-sapmchn   
 Semeru        |     | 17.0.2       | sem     |            | 17.0.2-sem          
               |     | 11.0.14      | sem     |            | 11.0.14-sem         
               |     | 8.0.322      | sem     |            | 8.0.322-sem         
 Temurin       |     | 17.0.2       | tem     |            | 17.0.2-tem          
               |     | 17.0.1       | tem     | local only | 17.0.1-tem          
               |     | 11.0.14      | tem     |            | 11.0.14-tem         
               |     | 8.0.322      | tem     |            | 8.0.322-tem         
 Trava         |     | 11.0.9       | trava   |            | 11.0.9-trava        
               |     | 8.0.232      | trava   |            | 8.0.232-trava       
 Zulu          |     | 17.0.2       | zulu    |            | 17.0.2-zulu         
               |     | 17.0.2.fx    | zulu    |            | 17.0.2.fx-zulu      
               |     | 11.0.14      | zulu    |            | 11.0.14-zulu        
               |     | 11.0.14.fx   | zulu    |            | 11.0.14.fx-zulu     
               |     | 8.0.322      | zulu    |            | 8.0.322-zulu        
               |     | 8.0.322.fx   | zulu    |            | 8.0.322.fx-zulu     
               |     | 7.0.332      | zulu    |            | 7.0.332-zulu        
               |     | 6.0.119      | zulu    |            | 6.0.119-zulu        
================================================================================

```

**Comment by den1k** [comment]
I'm not sure what the exact issue re: the JDK is. @huahaiy might have more insight but if you want to get it to run, try JDK 14.

**Comment by nkabir** [comment]
@den1k you are right. JDK14 works.

```
$ java --version
openjdk 14.0.2 2020-07-14
OpenJDK Runtime Environment Zulu14.29+23-CA (build 14.0.2+12)
OpenJDK 64-Bit Server VM Zulu14.29+23-CA (build 14.0.2+12, mixed mode, sharing)
```
I'm not sure if these are expected:

```
2022-02-14T16:23:27.264Z h-lpc-0 ERROR [datalevin.server:?] - 
               java.util.concurrent.ForkJoinWorkerThread.run  ForkJoinWorkerThread.java:  183
                 java.util.concurrent.ForkJoinPool.runWorker          ForkJoinPool.java: 1598
                      java.util.concurrent.ForkJoinPool.scan          ForkJoinPool.java: 1665
    java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec          ForkJoinPool.java: 1016
                    java.util.concurrent.ForkJoinTask.doExec          ForkJoinTask.java:  290
java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec          ForkJoinTask.java: 1429
                                                         ...                                 
                          datalevin.server/handle-read/fn/fn                 server.clj: 1645
                             datalevin.server/handle-message                 server.clj: 1621
                                    datalevin.server/tx-data                 server.clj: 1313
                                         datalevin.core/with                  core.cljc:  256
                                         datalevin.core/with                  core.cljc:  259
                               datalevin.db/transact-tx-data                    db.cljc:  959
                         datalevin.db/local-transact-tx-data                    db.cljc:  733
                      datalevin.db/local-transact-tx-data/fn                    db.cljc:  751
                                                         ...                                 
                             datalevin.entity/eval14041/fn/G                entity.cljc:  127 (repeats 2 times)
                               datalevin.entity.Entity/->txs                entity.cljc:  175
                                datalevin.entity/entity->txs                entity.cljc:  222
                             datalevin.entity/db->attr-types                entity.cljc:  204
                                 datalevin.db/eval13099/fn/G                    db.cljc:   46
                             clojure.core/-cache-protocol-fn           core_deftype.clj:  583
java.lang.IllegalArgumentException: No implementation of method: :-rschema of protocol: #'datalevin.db/IDB found for class: datalevin.remote.DatalogStore

```

But it does appear that all tests pass:

```
Testing pod.huahaiy.datalevin-test
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.lmdbjava.ByteBufferProxy$AbstractByteBufferProxy (file:/home/nkabir/.m2/repository/org/lmdbjava/lmdbjava/0.8.2/lmdbjava-0.8.2.jar) to field java.nio.Buffer.address
WARNING: Please consider reporting this to the maintainers of org.lmdbjava.ByteBufferProxy$AbstractByteBufferProxy
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release

Ran 256 tests containing 14938 assertions.
0 failures, 0 errors.
```
I'll see if @huahaiy has any suggestions on LTS development. Otherwise, I'll press ahead with JDK14. Thanks for the tip!

**Comment by huahaiy** [comment]
https://github.com/juji-io/datalevin/blob/master/project.clj#L58

I did not take the trouble to make these into lein profiles. We probably should, e.g. with profiles Java8, Java11, etc. PR welcome.

**Comment by nkabir** [comment]
From [this example](https://github.com/juji-io/datalevin/issues/94#issuecomment-1037580808),

I've narrowed down the behavior to this:

```
user=> (d/datoms @conn :ave)
[#datalevin/Datom [2 :person/born #inst "1971-01-01T00:00:00.000-00:00"]
 #datalevin/Datom [1 :person/born #inst "1969-01-01T00:00:00.000-00:00"]]

user=> (d/datoms @conn :eav 2 :person/born)
[#datalevin/Datom [2 :person/born #inst "1971-01-01T00:00:00.000-00:00"]]

user=> (d/datoms @conn :eav 1 :person/born)
[]
```

There's no query--just an attribute lookup by eid. I'm having trouble tracing this operation. If anyone happens to know where this lookup is implemented, please share. In the meantime, I'll keep looking and familiarizing myself with the codebase.

**Comment by ieugen** [comment]
@nkabir : if you push some code I might be able to take a look and help out.

**Comment by nkabir** [comment]
@ieugen I have not made any changes to the source. 

- forked the repository from @huahaiy 
- checked out tag `0.5.28`
- edited the `project.clj` and uncommented the settings for JDK11+ (I am using Clojure 1.10.3 and OpenJDK 64-Bit Server VM 17.0.1+12)
- I built the project and then ran some experiments against the build from @den1k [example](https://github.com/juji-io/datalevin/issues/94#issuecomment-1037580808) 

To eliminate DataScript as a cause (@den1k), I isolated the lookup to this operation:

```
user=> (d/datoms @conn :eav 1 :person/born)
[]
```

The source code appears to involve LMBD and Clojure macros so it's taking me some time to figure out what's going on since I'm not deeply familiar with either.

**Comment by huahaiy** [comment]
Fixed in 0.6.0

**Comment by nkabir** [comment]
Wonderful! Thank you!
